Universal emergency stop

IamPete1 opened the pull request:
This is the first bit of adding a universal E-stop for all vehicles. This sets any motor function to minimum pwm limit. I think this is the correct thing to do, also I have included k_starter and k_engine_run_enable as 'motors' as well as all the obvious ones, maybe there are some others that should also be included?

If this implementation is OK so far I will go through and replace the existing copter E-stop and add one to rover and plane.

khancyr commented:
Hello,
This won't work on rover or plane with reverse as zero is at trim and not min

rmackay9 commented:
I think it could also lead to I-term build up in the speed controllers on Rover because when it's done at this low level the motors library isn't aware and so it's not setting the throttle limit flags.

Pedals2Paddles commented:
Would the existing emergency stop (force disarm) accomplish this already?

peterbarker commented:
Needs to use overridable function calls here to accomplish what it needs to accomplish; Randy's point about integrator buildup seems to show that.  In that specific case the motors library could say `rover.rc().estopped()`, however.  Alternatively, there could be a (possibly pure) virtual callback the vehicles implement, (e.g. `RC_Channel_Rover::set_estop(bool)`) - but that probably leads to keeping the estopped state in multiple places.   OTOH, keeping that state in multiple places also means you could set the estop in multiple ways (not just RC).

The pattern I was after in RC_Channels was always having local methods in these switch statements - to keep those clean and to allow `override` to be tacked on trivially.

Naterater commented:
Could the E-Stop also stop the loops causing the integrator gain to build up, or do the EKF and control loops still need to function? Maybe a false mode switch to manual would prevent unwanted behavior after the E-stop is disengaged. Or maybe the vehicle should need to re-boot if the E-stop is triggered. Just ideas.

IamPete1 commented:
hopefully I have fixed some of the issues.

@khancyr your right, I have changed it to do the equivalent of set_output_scaled(0), I think this will be right in the case of reversing motors?

@rmackay9 I have added a relax function to rover to relax all the attitude controllers, do I also need to do this for the navigation controllers? I have also triggered the SRV_Channel estop in parallel with copters existing one, might catch some edge cases. 

@peterbarker I have changed to use the override function, I have gone with keeping the estop state at a vehicle level and also in the SRV_Channel library, makes it easier to integrate with copter's existing e-stop

next step is to extend to plane

IamPete1 commented:
have now added plane e-stop, works fine, however if the e-stop switch parameter is set it wont boot, do it  after boot and it works exactly as it should, not sure whats going on yet

magicrub commented:
Would be nice to have an option to mute outputs. Would be useful when this is failing the checks https://github.com/ArduPilot/ardupilot/issues/10465

peterbarker commented on ArduPlane/RC_Channel.cpp:
```diff
@@ -86,3 +86,23 @@ void RC_Channel_Plane::do_aux_function(const aux_func_t ch_option, const aux_swi
     }
 }
+
+void RC_Channel_Plane::do_aux_function_emergency_stop(const aux_switch_pos_t ch_flag)
```
This is the same across all vehicles; we should have a pure-virtual `set_emergency_stop` function implemented by all the vehicles.

rmackay9 commented on libraries/SRV_Channel/SRV_Channel.cpp:
```diff
@@ -181,9 +187,12 @@ uint16_t SRV_Channel::get_limit_pwm(LimitValue limit) const
+// return true if function is for a motor of any sort
 bool SRV_Channel::is_motor(SRV_Channel::Aux_servo_function_t function)
 {
-    return ((function >= SRV_Channel::k_motor1 && function <= SRV_Channel::k_motor8) ||
```
As discussed on the dev call, it would probably be best to revert this part of the change and instead create a new function to check if a function is for a "spinny bit".

rmackay9 commented:
As discussed on the dev call, there are a couple of things we should check:

- does this lead to I-term build-up for Copter?
- do the CAN ESCs work with this change?  I can check AP_ToshibaCAN ESCs

IamPete1 commented:
@rmackay9 For copter this just switches off the outputs at a servo level when the normal e-stop is triggered. So there is no change to the way e-stop works. Interestingly the motors library writes the pwm directly so actual bypasses this e-stop method. But this will catch other things such as the boost motor. 

I think CAN esc's should also be stopped because they read the pwm/the scaled output from the servo library before converting it to can. Would be great if you could double check.

IamPete1 commented:
This has passed travis, failed AppVeyor because of some unrelated thing but also failed semaphoreci, any ideas? Seems to be related to how I did the pure-virtual in RC_Channel

peterbarker commented:
```
[1658/1784] Compiling libraries/AP_Airspeed/examples/Airspeed/Airspeed.cpp
../../libraries/AC_PID/examples/AC_PID_test/AC_PID_test.cpp:34:52: error: invalid abstract type ‘RC_Channel_PIDTest’ for ‘obj_channels’
     RC_Channel_PIDTest obj_channels[NUM_RC_CHANNELS];
```

You need to update the example sketches.

try `./waf examples` if you want to build locally.

rmackay9 commented:
It would be nice if the RC_Channel's library had a get_emergency_stop() method so that we can remove the parallel state in Copter.

peterbarker commented:
@IamPete1 I think you could hear but not talk on today's devcall, so you generally know what was discussed...

IamPete1 commented:
This is now passing, I have fixed the example's thanks to Peter's tip. I will look at getting rid of the e-stop state in the vehicle codes. I think it makes seance to keep it in the SRV library rather than RC_Chan so you could potential do a e-stop by other means.  This will probably end up removing a bunch of stuff i have just added. But if i keep the commits separate then we can always come back to how it is now or If every ones happy i can squash.

IamPete1 commented:
@rmackay9 Is this now more what you had in mind? I think its better than it was also brings e-stop logging to plane and rover.

rmackay9 commented:
as discussed on the dev call, moving the state to the SRV_Channel library sounds good.

IamPete1 commented:
this is now squashed and tided i think it just needs merging

rmackay9 commented on libraries/RC_Channel/RC_Channel.cpp:
```diff
@@ -442,6 +442,7 @@ void RC_Channel::init_aux_function(const aux_func_t ch_option, const aux_switch_
     case CLEAR_WP:
     case COMPASS_LEARN:
     case LANDING_GEAR:
+    case MOTOR_ESTOP:
```
I think this is a change in behaviour.  We should be initialising the e-stop state based on the switch's position at startup.

IamPete1 replied:
your right, good catch, I also left it in the copter init function, will fix this evening

IamPete1 replied:
fixed

rmackay9 commented on APMrover2/APMrover2.cpp:
```diff
@@ -320,6 +320,12 @@ void Rover::update_GPS(void)
 
 void Rover::update_current_mode(void)
 {
+    // check for emergency stop
```
sorry to be annoying but can we move this to Steering.cpp's Rover::set_servos() function?

rmackay9 commented:
Sorry, i found one more small thing above re initialisation of the switch.

IamPete1 commented:
I think the only things missing now from rover and plane is the auto disarm after a set time of e-stop. For plane  there doesn't seem to be a speed integrator that would directly cause a throttle spike when coming out of e-stop, it seems to be the energy error in the TECS i think its probably not a good idea to zero that. Maybe some with more experience with it can give me some advice for a follow up PR.

rmackay9 commented:
Merged, thanks!
